python/contrib/dag/__init__.py

```diff
@@ -23,8 +23,16 @@ class DAG(object):
 
     def __init__(self):
         """ Construct a new DAG with no nodes or edges. """
+        self.__original_graph = None
         self.reset_graph()
 
+    # Added by the MOOSE group
+    def __createOriginalGraph(self):
+        """ Private method to create and protect the current state of the graph """
```

Comment:
Hmm... Not sure this method always works. If you've started deleting nodes, it would give you a clone of the current state. Also, 'protect' really isn't true. Since you are returning the original graph (not a deep copy). The user could also change the cloned graph.

I think I remember you talking about the need for this method. You needed to be able to tell the graph that you were done adding nodes and edges and needed it to save it's current state. You might just call this method clone(). More work is needed if you really want this method to do the right thing when the user screws up.

Reply:
Yeah... I know its not really protected. This could be better designed. This was a hasty addition.